addon/devtoolsPanel.js

```diff
@@ -1,47 +1,232 @@
+                console.error(e);
+            });
+    }
+
+    function addPathCell(row, label) {
+        const td = document.createElement('td');
+        td.classList.add('path');
+        if (label.inspected) {
+            td.appendChild(document.createTextNode(label.path));
+        } else {
+            let a = document.createElement('a');
+            a.href = '#';
+            a.dataset.path = label.path;
+            a.addEventListener('click', inspectLabeled);
+            a.appendChild(document.createTextNode(label.path));
+            td.appendChild(a);
+        }
+        row.appendChild(td);
+    }
+
+    function updateLabel(event) {
+        backgroundPort.postMessage({
+            type: 'label',
+            tabId: browser.devtools.inspectedWindow.tabId,
+            selector: event.target.dataset.path,
+            label: event.target.value.trim(),
+        });
+    }
+
+    function addLabelCell(row, label) {
+        const td = document.createElement('td');
+        td.classList.add('label');
+        const input = document.createElement('input');
+        input.setAttribute('placeholder', 'unlabeled');
+        input.value = label.label;
+        input.dataset.path = label.path;
+        input.addEventListener('change', updateLabel);
+        td.appendChild(input);
+        row.appendChild(td);
+    }
+
+    function removeLabeled(event) {
+        event.preventDefault();
+        backgroundPort.postMessage({
+            type: 'label',
+            tabId: browser.devtools.inspectedWindow.tabId,
+            selector: event.target.dataset.path,
+            label: '',
+        });
+    }
+
+    function addClearLabelCell(row, label) {
+        const td = document.createElement('td');
+        td.classList.add('action');
+        const a = document.createElement('a');
+        a.href = '#';
+        a.classList.add('action-button');
+        a.dataset.path = label.path;
+        a.dataset.label = label.label;
+        a.setAttribute('title', label.inspected ? 'Clear Label' : 'Remove Label');
+        a.addEventListener('click', removeLabeled);
+        a.appendChild(document.createTextNode(label.inspected ? '⦸' : '✖'));
+        td.appendChild(a);
+        row.appendChild(td);
+    }
+
+    const table = document.createElement('table');
+    const headerRow = document.createElement('tr');
+    addTextCell(headerRow, 'Selector', true).classList.add('path');
+    addTextCell(headerRow, 'Element', true).classList.add('preview');
+    addTextCell(headerRow, 'Label', true).classList.add('label');
+    addTextCell(headerRow, '').classList.add('action');
+    table.appendChild(headerRow);
+
+    for (const label of labeled) {
+        const row = document.createElement('tr');
+        row.classList.add('hover');
+
+        addPathCell(row, label);
+        addTextCell(row, label.preview).classList.add('preview');
+        addLabelCell(row, label);
+        addClearLabelCell(row, label);
+
+        table.appendChild(row);
```

Comment:
It'd be totally killer if hovering a table row highlit the corresponding element. Just in a short playtest, I found I was missing that. Firefox's Inspector doesn't do a very good job standing in for this: clicking the Selector link, while it does select the element in the Inspector, unhighlights it the moment you move the mouse another pixel. And your hand is already on the mouse. Also, if you're just trying to get your bearings, having to continually re-click the Fathom devpane to hover another row is a drag.

Doesn't block merge, of course, but we should file this unless you can think of why it's a bad idea. It's so great to have an actual UI to polish! :-D

Reply:
this is probably going to be deferred.

the current highligher needs some tweaks, especially because of how it interacts poorly with page scrolling and changes.  perhaps a better approach would be to use pure css to style the element appropriately (eg. using an `outline`) and then all we'd need to to is add/remove classes to the current element to show it.